build: remove deps on legacy nodejs rules rollup_bundle internals

gregmagolan opened the pull request:
The legacy nodejs rules rollup_bundle is now deprecated and will be removed in the nodejs rules 1.0 release due in mid-November. This PR brings in the rules_nodejs internal API deps that ng_rollup_bundle, ng_package and ls_rollup_bundle depend on into this repo to break the dependency. In the future these rules should switch to use the new rollup_bundle via a macro as done in https://github.com/angular/angular/pull/33329 but this is not possible right now due to the complication of having esm5 re-rooted ts_library dependencies.

Note: repo has been updated to the newest working version of rollup which is 1.25.2. There is some regression in 1.26.0 which causes bundling failures of the following sort:

```
ERROR: /Users/greg/google/angular/packages/localize/BUILD.bazel:20:1: Optimizing JavaScript packages/localize/localize.umd.min.js [terser] failed (Exit 1) terser.sh failed: error executing command bazel-out/host/bin/external/npm/terser/bin/terser.sh bazel-out/darwin-fastbuild/bin/packages/localize/localize.umd.js --output bazel-out/darwin-fastbuild/bin/packages/localize/localize.umd.min.js ... (remaining 5 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
Parse error at packages/localize/localize.umd.js:491,4
    export * from './src/constants';
    ^
ERROR: Export statement may only appear at top level
    at js_error (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/parse.js:357:11)
    at Dn.visit (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/scope.js:347:13)
    at Dn._visit (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:1207:24)
    at AST_Export._walk (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:718:17)
    at walk_body (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:168:17)
    at AST_Function.call (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:430:13)
    at descend (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:1208:21)
    at Dn.visit (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/scope.js:256:13)
    at Dn._visit (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:1207:24)
    at AST_Function._walk (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:424:24)
[Function]
Target //packages/localize:npm_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /Users/greg/google/angular/packages/localize/BUILD.bazel:20:1 Optimizing JavaScript packages/localize/localize.umd.min.js [terser] failed (Exit 1) terser.sh failed: error executing command bazel-out/host/bin/external/npm/terser/bin/terser.sh bazel-out/darwin-fastbuild/bin/packages/localize/localize.umd.js --output bazel-out/darwin-fastbuild/bin/packages/localize/localize.umd.min.js ... (remaining 5 argument(s) skipped)
```

Will leave that for another day.

Terser also updated to 4.3.3. Updating to 4.3.4 (https://github.com/terser/terser/blob/master/CHANGELOG.md) turns comments preservation on by default which increases the size of the `//packages/core/test/bundling/todo:bundle.min.js` in CI. After https://github.com/bazelbuild/rules_nodejs/pull/1317 terser can be updated to the latest as passing `—comments /a^/` to args can turn off all comments for the `//packages/core/test/bundling/todo:bundle.min.js` size test.

alexeagle commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -102,12 +110,152 @@ WELL_KNOWN_GLOBALS = {p: _global_name(p) for p in [
+
+    return depset(direct = rerooted_files)
+
+def _terser(ctx, input, output):
```
why do we need our own implementation? can't we get this from `@npm_bazel_terser`?

gregmagolan replied:
`@npm_bazel_terser` implementation more complicated with the directory & sourcemap support than needed and also accesses ctx.attr.debug which ng_package does not have. Plus access internals of rules has caused us problems in the past. My plan is to deconstruct this into a macro after moving it to rules_nodejs so the bundling, minifying & packaging steps are done by separate rules. This rule does both right now still accesses the npm_package internals here.

alexeagle commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -102,12 +110,152 @@ WELL_KNOWN_GLOBALS = {p: _global_name(p) for p in [
+
+    return map_output
+
+def _compute_node_modules_root(ctx):
```
also couldn't re-use this?

gregmagolan replied:
You mean re-use in ng_rollup_bundle? Avoided that as ng_package is moving to rules_nodejs shortly.

alexeagle commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -102,12 +110,152 @@ WELL_KNOWN_GLOBALS = {p: _global_name(p) for p in [
+        node_modules_root = "external/npm/node_modules"
+    return node_modules_root
+
+def _write_rollup_config(ctx, root_dir, filename = "_%s.rollup.conf.js"):
```
how is it different from the one in `@npm_bazel_rollup`?

gregmagolan replied:
its still templated to support the `function resolveBazel(` which is needed for the esm5 inputs. Hopefully this can get a massive cleanup after moving to rules_nodejs and it can just use the new rollup_bundle and maybe use babel to downlevel to esm5 from ts_library es2015 output.

alexeagle commented on tools/ng_rollup_bundle/ng_rollup_bundle.bzl:
```diff
@@ -0,0 +1,456 @@
+
+"""Rollup with Build Optimizer
+
+   This provides a variant of the [rollup_bundle] rule that works better for Angular apps.
```
should there be a TODO to boil this down when esm5 support is dropped? Igor says that's coming in 10 so we could land it in about three months

gregmagolan replied:
👍

alexeagle commented on tools/ng_rollup_bundle/ng_rollup_bundle.bzl:
```diff
@@ -0,0 +1,456 @@
+    # The format is actually AMD and not UMD, but we are afraid to rename
+    # the file because that would likely break the IDE and other integrations that
+    # have the path hardcoded in them.
+    ng_rollup_bundle(
```
how did you test it? build LS package before and after the change to see we produced identical bundles?

gregmagolan replied:
yes; I diffed them and they were identical character for character. I'll try generating the bundle with es2015 inputs using the new rollup_bundle and then downleving to es5 with babel and see if it is still identical; then ls_rollup_bundle won't depend on esm5 at least.

gregmagolan replied:
bundle differs quite a bit with esm2015+babel. will leave the esm2015+babel change out of this https://github.com/angular/angular/pull/33414

gregmagolan commented:
Some notes on this:

After discussing with @alexeagle I moved ng_rollup_bundle out of the `@angular/bazel` package and into `/tools/ng_rollup_bundle` as it is not used externally at all AFAICT and `@angular/bazel` package is most likely going to be move to rules_nodejs soon.

Ideally, `ng_rollup_bundle` would be a macro that uses the new rollup_bundle rule, terser_minified & npm_package_bin to run brotli but because of its esm5 requirement it means we need re-run tsc to generated esm5 with the `esm5_outputs_aspect` and then re-root the tree with `flatten_esm5`

I thought about making an esm5_sources rule and use the output of that with the new rollup_bundle but the esm5 input tree in that shape is not simple to use with the new rollup_bundle. Also, Igor says that esm5 is being dropped for Angular 10 so it would not be worth the effort as we can switch to the new rollup_bundle in about 3 months.

`ng_rollup_bundle` still depends on `load("@build_bazel_rules_nodejs//internal/common:module_mappings.bzl", "get_module_mappings")` but all of its deps on the legacy rollup_bundle has been removed. The `module_mappings.bzl` dep will go away when we can switch to the new rollup_bundle after esm5 inputs are no longer needed.

gregmagolan commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -102,12 +110,152 @@ WELL_KNOWN_GLOBALS = {p: _global_name(p) for p in [
 _DEPSET_TYPE = "depset"
 
-def _rollup(ctx, bundle_name, rollup_config, entry_point, inputs, js_output, format = "es", module_name = "", include_tslib = False):
+def _collect_es6_sources(ctx):
```
I tried that originally and it blew up but that may have been due to other issues. I'll try it again here as es6 sources should work without re-rooting. The re-rooting was originally due to the legacy `.clusure.js` naming.

gregmagolan replied:
_collect_es6_sources() removed and replaced with access to JSEcmaScriptModuleInfo

gregmagolan commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -102,12 +110,152 @@ WELL_KNOWN_GLOBALS = {p: _global_name(p) for p in [
+        if (path.startswith("../")):
+            path = "external/" + path[3:]
+
+        rerooted_file = ctx.actions.declare_file(
```
No collision. Had trouble getting to work without re-rooting so limited the changes in the PR to get to green. I'll try again now that its all green and the change is isolated.

gregmagolan replied:
updated this so es2015 sources are no longer re-rooted; JSEcmaScriptModuleInfo used instead; some changes to ng_package.ts needed to handle the new paths

gregmagolan commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -38,6 +31,21 @@ def _debug(vars, *args):
+_DEFAULT_ROLLUP = "@npm//rollup/bin:rollup"
+_DEFAULT_TERSER = "@npm//terser/bin:terser"
+
+def _rollup_module_mappings_aspect_impl(target, ctx):
```
I'm not sure why it was like that in the legacy rollup_bundfle. Slightly different implementation of the aspect in module_mappings.bzl so there may be a reason?
```
# When building a mapping for use at runtime, we need paths to be relative to
# the runfiles directory. This requires the workspace_name to be prefixed on
# each module root.
def _module_mappings_runtime_aspect_impl(target, ctx):
    if target.label.workspace_root:
        # We need the workspace_name for the target being visited.
        # Skylark doesn't have this - instead they have a workspace_root
        # which looks like "external/repo_name" - so grab the second path segment.
        # TODO(alexeagle): investigate a better way to get the workspace name
        workspace_name = target.label.workspace_root.split("/")[1]
    else:
        workspace_name = ctx.workspace_name
    mappings = get_module_mappings(
        target.label,
        ctx.rule.attr,
        workspace_name = workspace_name,
        mappings_attr = "runfiles_module_mappings",
    )
    return struct(runfiles_module_mappings = mappings)

module_mappings_runtime_aspect = aspect(
    _module_mappings_runtime_aspect_impl,
    attr_aspects = _MODULE_MAPPINGS_DEPS_NAMES,
)
```
I'll investigate

gregmagolan commented on tools/ng_rollup_bundle/ng_rollup_bundle.bzl:
```diff
@@ -0,0 +1,456 @@
+
+[Rollup]: https://rollupjs.org/
+[Build Optimizer]: https://www.npmjs.com/package/@angular-devkit/build-optimizer
+[rollup_bundle]: https://bazelbuild.github.io/rules_nodejs/rollup/rollup_bundle.html
```
👍

gregmagolan commented:
`ng_package` no longer depends on any legacy rollup_bunde internals but it does still depend on `internal/common/module_mappings.bzl` & npm_package internals:

```
load("@build_bazel_rules_nodejs//internal/common:module_mappings.bzl", "get_module_mappings")
load(
    "@build_bazel_rules_nodejs//internal/npm_package:npm_package.bzl",
    "NPM_PACKAGE_ATTRS",
    "NPM_PACKAGE_OUTPUTS",
    "create_package",
)
```
My plan is to deconstruct this into a macro after moving it to rules_nodejs so the bundling, minifying & packaging steps are done by separate rules. This rule does both right now still accesses the npm_package internals here.

gregmagolan commented:
Split this PR up into a number of different PRs so they are easier to review & land:

https://github.com/angular/angular/pull/33426
https://github.com/angular/angular/pull/33433
https://github.com/angular/angular/pull/33434
https://github.com/angular/angular/pull/33435

I also removed the collect_es6_sources() function and used the new `JSEcmaScriptModuleInfo` aspect to gather these sources as they no longer need to be re-rooted since they have `.mjs` extension and are no longer end in `.closure.js`.

I also removed the dep on `internal/common/module_mappings.bzl` so we have fewer deps on rules_nodejs internals. The only ones left are `internal/web_package/web_package.bzl` in the @angular/bazel schematic and `internal/npm_package` in ng_package.

gregmagolan commented:
Having some trouble with the rollup version to use in the repo as updating it seems to break some of the benchmarks here. Investigating...

gregmagolan committed:
build: remove deps on legacy nodejs rules rollup_bundle internals

The legacy nodejs rules rollup_bundle is now deprecated and will be removed in the nodejs rules 1.0 release due in mid-November. This PR brings in the rules_nodejs internal API deps that ng_rollup_bundle, ng_package and ls_rollup_bundle depend on into this repo to break the dependency. In the future these rules should switch to use the new rollup_bundle via a macro as done in https://github.com/angular/angular/pull/33329 but this is not possible right now due to the complication of having esm5 re-rooted ts_library dependencies.

The es6 sources now have .mjs extensions so they no longer need to be re-rooted to `{package}.es6`. This eliminates the need for the collect_es6_sources() function.

Note: repo has been updated to the newest working version of rollup which is 1.25.2. There is some regression in 1.26.0 which causes the following bundling failure:

```
ERROR: /Users/greg/google/angular/packages/localize/BUILD.bazel:20:1: Optimizing JavaScript packages/localize/localize.umd.min.js [terser] failed (Exit 1) terser.sh failed: error executing command bazel-out/host/bin/external/npm/terser/bin/terser.sh bazel-out/darwin-fastbuild/bin/packages/localize/localize.umd.js --output bazel-out/darwin-fastbuild/bin/packages/localize/localize.umd.min.js ... (remaining 5 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
Parse error at packages/localize/localize.umd.js:491,4
    export * from './src/constants';
    ^
ERROR: Export statement may only appear at top level
    at js_error (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/parse.js:357:11)
    at Dn.visit (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/scope.js:347:13)
    at Dn._visit (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:1207:24)
    at AST_Export._walk (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:718:17)
    at walk_body (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:168:17)
    at AST_Function.call (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:430:13)
    at descend (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:1208:21)
    at Dn.visit (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/scope.js:256:13)
    at Dn._visit (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:1207:24)
    at AST_Function._walk (/private/var/tmp/_bazel_greg/5e8f8a9cd1c6fbc1afd11e37ee1fe2e5/sandbox/darwin-sandbox/79/execroot/angular/bazel-out/host/bin/external/npm/terser/bin/terser.sh.runfiles/npm/node_modules/terser/lib/ast.js:424:24)
[Function]
Target //packages/localize:npm_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /Users/greg/google/angular/packages/localize/BUILD.bazel:20:1 Optimizing JavaScript packages/localize/localize.umd.min.js [terser] failed (Exit 1) terser.sh failed: error executing command bazel-out/host/bin/external/npm/terser/bin/terser.sh bazel-out/darwin-fastbuild/bin/packages/localize/localize.umd.js --output bazel-out/darwin-fastbuild/bin/packages/localize/localize.umd.min.js ... (remaining 5 argument(s) skipped)
```

Will leave that for another day.

Terser also updated to 4.3.3. Updating to 4.3.4 (https://github.com/terser/terser/blob/master/CHANGELOG.md) turns comments preservation on by default which increases the size of the //packages/core/test/bundling/todo:bundle.min.js in CI. After bazelbuild/rules_nodejs#1317 terser can be updated to the latest as passing —comments /a^/ to args can turn off all comments for the //packages/core/test/bundling/todo:bundle.min.js size test.

gregmagolan commented:
Bundle size difference may be that the newer rollup version is better at preserving comments? Reviews: see ng_package example package golden file in diff. Comments that were dropped with previous rollup bundle version were preserved now.

gregmagolan commented on packages/bazel/src/ng_package/BUILD.bazel:
```diff
@@ -27,9 +27,30 @@ nodejs_binary(
+    ],
+    entry_point = "@npm//:node_modules/rollup/dist/bin/rollup",
+    install_source_map_support = False,
+    visibility = ["//visibility:public"],
```
I think it needs to be available to all packages as its used in different context

```
_DEFAULT_ROLLUP = "@npm_angular_bazel//src/ng_package:rollup_for_ng_package"
```

for users of the published `@angular/bazel` package and

```
_INTERNAL_NG_PACKAGE_DEFAULT_ROLLUP = "//packages/bazel/src/ng_package:rollup_for_ng_package"
```

in the angular repo directly.

But the default visibility on the package is public so no need to explicitly set it on the target. Removed that line.

IgorMinar replied:
ok. thanks

AndrewKushnir commented:
@gregmagolan could you please check if this change should be included only into master branch or  in patch branch as well (since we just created a new patch branch - `9.0.x`)? Thank you.

gregmagolan commented:
@AndrewKushnir Ideally master & patch as we want to keep compat between @angular/bazel 9.0.x and rules_nodejs 1.0.0. Should merge cleanly as the branch is now.

AndrewKushnir commented:
Thanks @gregmagolan 👍 I've removed the "cleanup" label.
